### DISCUSSION TITLE
mario.py의 copy에 대한 고찰

### DISCUSSION THREAD

**Discussion by right-chan** [N1-314/Intrinsic-Curiosity-Module---PyTorch#7] [Show and tell]
- [verification_env_obs_id.ipynb](https://github.com/N1-314/Intrinsic-Curiosity-Module---PyTorch/blob/96a54c4124bbd43815e3b025fffbebe4a1871aa8/verification_env_obs_id.ipynb) 을 보면 state의 id를 확인할 수 있습니다.

- `ResizeObservation`에서 state가 copy되는 것이 확인되었습니다. 

따라서 `mario.py`에서 Resize 전에 이미지를 저장하는 `RecordFrames`는 copy가 필요하고, `Resize`후에 이미지를 저장하는 `SkipStackObservation`는 copy가 필요 없습니다.

**Comment by HyeAnn-Lee** [comment]
똑똑하신 고찰입니다. 그럼 `gym.wrappers.ResizeObservation` wrapper를 사용하지 않는다면 (= resize하기 싫고 그냥 원상태 그대로 training하고 싶어한다면) skipstack에서 다시 `copy()`가 필요한걸까요? 나중에 출근해서 확인해보겠습니다.

**Reply by right-chan** [reply]
추가로 확인한 결과 `ResizeObservation` wrapper뿐만 아니라 `GrayScaleObservation` wrapper역시 state가 copy되는 것을 확인했습니다.

`cv2`모듈이 메모리를 새롭게 할당하는 것 같아요.

따라서 env에 `ResizeObservation`, `GrayScaleObservation` wrapper를 적용하지 않았다면, `SkipStackObservation`에서 `copy()`가 필요합니다.
